s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java

```diff
@@ -224,10 +224,21 @@ public void doAttach() throws Exception {
           injectType( FileType.IMAGINARY );
           return;
         }
+
+        // One common case is "403 Access Denied", for bucket names which exist in the same region
```

Comment:
@nawaz34-hitachivantara I don't understand the business logic in the different handling between 403 error and 404 error. What is the ultimate goal of both cases. I'm not familiar with this code.

- 404 "not found" still returns with but a file type of imaginary
- 403 "access denied" throws an error


Correct me if I'm interpreting the code wrong?

Reply:
Before this change, a bucket of same name on another AWS account would be recognized as a normal, existing folder, despite the 403 error received in this method. Code would later fail when trying to do something with it, as access is denied.

Recognizing the other-account-bucket as imaginary, would mean it simply considered not found. This misattribution causes issues elsewhere. For example, when testing the root path of an S3 connection, using the name of a bucket from another account would be considered valid, despite trying to read anything from that bucket effectively failing.

Throwing an exception seems to be the best approach to follow here. This is how Apache VFS APIs seem to handle no access issues. This is also how effectively, `handleAttachExceptionFallback`, handles the issue for regular non-bucket files.

This is explained in the comment somewhat.